### PR TITLE
Convert identity API to Typescript

### PR DESCRIPTION
**Pull request by mxdvl**
## What does this change?

Converts the [Identity API](https://github.com/guardian/frontend/blob/typescript-identity-api/static/src/javascripts/projects/common/modules/identity/api.ts) to Typescript.

- Update the shape of getUserFromCookie response to match a IDAPI response:
  - `displayName` is now a child of `publicFields`
- Adds types based of typical responses and available information. This might be the area that 

I have also provided a rough type over `config.get` to make Typescript happy, but it does not feel like the proper way to do this. I’ll put a warning sign to bring attention to this: ⚠️

## Does this change need to be reproduced in dotcom-rendering ?

- [X] No
- [ ] Yes (please indicate your plans for DCR Implementation)

## What is the value of this and can you measure success?

Getting a step closer to removing `ajax`, `reqwest` and `lib/fetch` from our codebase.

### Does this change break ad-free?

- [X] No
- [ ] It did, but tests caught it and I fixed it
- [ ] It did, but there was no test coverage so I added that then fixed it

### Tested

- [ ] Locally
- [ ] On CODE (optional)

### CODE REVIEW

File: static/src/javascripts/projects/common/modules/identity/api.ts [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+		method: 'PATCH',
+		credentials: 'include',
+		mode: 'cors',
+		body: JSON.stringify(consents),
+	}).then((resp) => {
+		if (resp.ok) return Promise.resolve();
+		return Promise.reject();
+	});
+
+export const ajaxSignIn = (
```
Comment (dominickendrick) [inline static/src/javascripts/projects/common/modules/identity/api.ts:397]: I think Identity has deprecated `smartlock / browser credentials` support https://github.com/guardian/identity-frontend/pull/607

File: static/src/javascripts/projects/common/modules/identity/api.ts [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+	userGroups: UserGroups[];
+	publicFields: {
+		username: string;
+		displayName: string;
+	};
+	statusFields: {
+		userEmailValidated: boolean;
+	};
+	privateFields: {
+		brazeUuid: string;
```
Comment (dominickendrick) [inline static/src/javascripts/projects/common/modules/identity/api.ts:67]: I also have `hasRepermissioned: true` as an attribute on `privateFeilds`

File: static/src/javascripts/projects/common/modules/identity/api.ts [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+export type IdentityUser = {
+	dates: { accountCreatedDate: string };
+	consents: UserConsents[];
+	userGroups: UserGroups[];
+	publicFields: {
+		username: string;
+		displayName: string;
+	};
+	statusFields: {
+		userEmailValidated: boolean;
```
Comment (dominickendrick) [inline static/src/javascripts/projects/common/modules/identity/api.ts:64]: I have a `hasRepermissioned` attribute with a `boolean` type value in `statusFields`

File: static/src/javascripts/projects/common/modules/identity/api.ts [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+const signOutCookieName = 'GU_SO';
+const fbCheckKey = 'gu.id.nextFbCheck';
+let idApiRoot = '';
+let profileRoot = '';
+
+export const init = (): void => {
+	idApiRoot = config.get('page.idApiUrl');
+	mediator.emit('module:identity:api:loaded');
+	profileRoot = config.get('page.idUrl');
+};
```
Comment (dominickendrick) [inline static/src/javascripts/projects/common/modules/identity/api.ts:108]: This feels like a bit of an anti-pattern (and probably a legacy code hangover) - a callee has to run `init` before they call any other functions that utilise `idApiRoot` or `profileRoot` exported from this module. 
Could we just define these as `const` when the module is first imported ie
```
const idApiRoot = config.get('page.idApiUrl');
const profileRoot = config.get('page.idUrl');
mediator.emit('module:identity:api:loaded');
```
and drop the whole `init` function 🤔

File: static/src/javascripts/projects/common/modules/identity/api.ts [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+
+export const decodeBase64 = (str: string): string =>
```
Comment (dominickendrick) [inline static/src/javascripts/projects/common/modules/identity/api.ts:110]: It looks like this function is defined and used in another part of the codebase toohttps://github.com/guardian/frontend/search?q=decodeBase64 - is it worth moving it to a util somewhere ?

File: static/src/javascripts/projects/common/modules/identity/api.ts (hunk 4 above)
Comment (mxdvl) [reply static/src/javascripts/projects/common/modules/identity/api.ts]: Fixed in 7edb4ef

File: static/src/javascripts/projects/common/modules/identity/api.ts (hunk 3 above)
Comment (mxdvl) [reply static/src/javascripts/projects/common/modules/identity/api.ts:64]: Added in d010572ce4

File: static/src/javascripts/projects/common/modules/identity/api.ts (hunk 5 above)
Comment (mxdvl) [reply static/src/javascripts/projects/common/modules/identity/api.ts:106]: Actually, for Javascript, it’s only<sup>1</sup> in the Identity API–and its archived flow files: https://github.com/guardian/frontend/search?l=JavaScript&q=decodeBase64

<sup>1</sup>: _I feel like https://github.com/guardian/frontend/blob/5b970cd7308175cfc1bcae2d4fb8c06ee13c5fa0/common/app/templates/inlineJS/nonBlocking/getUserData.scala.js is a special case._

File: static/src/javascripts/projects/common/modules/identity/api.ts (hunk 1 above)
Comment (mxdvl) [reply static/src/javascripts/projects/common/modules/identity/api.ts:395]: I’m afraid I can’t see that repo, but happy to remove this bit of code.

File: static/src/javascripts/projects/common/modules/identity/api.ts (hunk 5 above)
Comment (dominickendrick) [reply static/src/javascripts/projects/common/modules/identity/api.ts:100]: ah cool, I'm happy to leave it in then : )

**Review by dominickendrick — APPROVED** [review]
Looks great, thanks for doing this!

**Pull request merged**